File: cycle.js
function has_cycle(head) {
  const seen = new Set()
  let curr = head

  while (true) {
    if (!curr) {
      break
    } else if (!seen.has(curr)) {
      seen.add(curr)
      curr = curr.next
    } else {
      return true
    }
  }

  return false
}

File: cycle.cpp
#include <set>

bool has_cycle(Node *head) {
  std::set<Node *> seen;
  Node *curr = head;

  while (true) {
    if (curr == NULL) {
      break;
    }
    else if (seen.count(curr) == 0) {
      seen.insert(curr);
      curr = curr->next;
    } else {
      return true;
    }
  }
  
  return false;
}